java/server/src/org/openqa/grid/web/servlet/handler/WebDriverRequest.java

```diff
@@ -65,17 +73,46 @@ public ExternalSessionKey extractSession() {
       JsonObject map = new JsonParser().parse(json).getAsJsonObject();
       // Current W3C has required / desired capabilities wrapped in a 'capabilities' object.
       // This will need to be updated if/when https://github.com/w3c/webdriver/pull/327 gets merged
-      if (map.has("capabilities")) {
-        JsonObject outerCapabilities = map.getAsJsonObject("capabilities");
-        if (outerCapabilities.has("desiredCapabilities")) {
-          JsonObject desiredCapabilities = outerCapabilities.getAsJsonObject("desiredCapabilities");
+      if (map.has(CAPABILITIES)) {
+        JsonObject outerCapabilities = map.getAsJsonObject(CAPABILITIES);
+        if (outerCapabilities.has(DESIRED_CAPABILITIES)) {
+          JsonObject desiredCapabilities = outerCapabilities.getAsJsonObject(DESIRED_CAPABILITIES);
           return new JsonToBeanConverter().convert(Map.class, desiredCapabilities);
         }
       }
-      JsonObject dc = map.get("desiredCapabilities").getAsJsonObject();
+      JsonObject dc = map.get(DESIRED_CAPABILITIES).getAsJsonObject();
       return new JsonToBeanConverter().convert(Map.class, dc);
+
     } catch (Exception e) {
       throw new GridException("Cannot extract a capabilities from the request: " + json, e);
     }
   }
+
+  @Override
+  public String getBody() {
+    String json =  super.getBody();
+    try {
+      Map<String, Object> capsMap = getDesiredCapabilities();
+      if (capsMap == null) {
+        return json;
+      }
+      JsonObject map = new JsonParser().parse(json).getAsJsonObject();
+      JsonElement dc = new BeanToJsonConverter().convertObject(capsMap);
+      if (map.has(CAPABILITIES)) {
+          map.getAsJsonObject(CAPABILITIES)
+              .add(DESIRED_CAPABILITIES, new BeanToJsonConverter().convertObject(dc));
+      } else {
+        map.add(DESIRED_CAPABILITIES, dc);
+      }
+      return new Gson().toJson(map);
+    } catch (Exception e) {
+      return json;
+    }
+  }
+
+  @Override
+  public ServletInputStream getInputStream() throws IOException {
```

Comment:
Took care of this by moving the call to `getBody()` from within the SeleniumBasedRequest.getInputStream()`. Chose to do `setBody(getBody());` instead of 
`return new ServletInputStreamImpl(new ByteArrayInputStream(body.getBytes()));` to avoid repetition of the `getBytes()` invocation.